pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -66,10 +66,29 @@ This file implements string parsing and creation for NumPy datetime.
  *
  * Returns 0 on success, -1 on failure.
  */
+
+#define FORMAT_STARTSWITH(ch) \
+  if (exact) { \
+    if (!format_len || *format != ch) { \
+        goto parse_error; \
+    } \
+    ++format; \
+    --format_len; \
+  } else { \
+    if (format_len > 0) { \
+        if (*format != ch) { \
+            goto parse_error;  \
+        } \
+        ++format; \
+        --format_len; \
+    } \
+  } \
```

Comment:
Nit: this could probably be simplified to something like this:
```c++
if (*format != ch) goto parse_error;
if (exact ^ format_len) goto parse_error;
++format;
--format_len;
```